### ISSUE TITLE
element 0 of tensors does not require grad and does not have a grad_fn

### ISSUE THREAD

**Issue by sosoIrving** [frgfm/torch-cam#26]
I get this error when I use the API.

`model = models.resnet34(pretrained=True).eval()
cam = GradCAMpp(model,'layer4')
import numpy as np
from PIL import Image
from torch.autograd import Variable

from torchvision.transforms.functional import normalize,resize,to_tensor
img_path = ''
pil_img = Image.open(img_path)
#pil_img = np.expand_dims(img_tensor,axis=0)

img_tensor = normalize(to_tensor(resize(pil_img,(224,224))),[0.485,0.456,0.406],[0.229,0.224,0.225])
_ = _forward(model,torch.stack((img_tensor,img_tensor)))
scores = _forward(model, img_tensor.unsqueeze(0))
#img_variable = Variable(img_tensor.unsqueeze(0),requires_grad=True)
cam(class_idx=1,scores=scores)
`

**Comment by frgfm** [comment]
Hi @sosoIrving,

Could you edit your issue description to display the error? I'm not sure whether this is a typo but the error you mentioned is not in it. 

Also, I see that you use a function called `_forward` whose definition is not here. I need to be able to reproduce your error for me to be able to help :sweat_smile: Please ensure that the snippet you provide is raising the error as a standalone script :pray: 

Thanks!

**Comment by lm-b** [comment]
Hi, 

I had the same error based following the example code snippets in the docs. For GradCAM, this has

```
>>> from torchvision.models import resnet18
>>> from torchcam.cams import GradCAM
>>> model = resnet18(pretrained=True).eval()
>>> cam = GradCAM(model, 'layer4')
>>> with torch.no_grad(): scores = model(input_tensor)
>>> cam(class_idx=100, scores=scores)
```

I'm guessing based on your comment above, @frgfm that we should just use `scores=model(input_tensor)` rather than the line `with torch.no_grad(): scores = model(input_tensor)` ?

**Comment by sosoIrving** [comment]
> 
> 
> Hi @sosoIrving,
> 
> Could you edit your issue description to display the error? I'm not sure whether this is a typo but the error you mentioned is not in it. If the error that you get is the one in the title, then it looks like the problem comes most likely from your forward. Gradient based CAM methods need to be able to backprop through the architecture. If you use for instance `with torch.no_grad()` then this will not work.
> 
> Also, I see that you use a function called `_forward` whose definition is not here. I need to be able to reproduce your error for me to be able to help 😅 Please ensure that the snippet you provide is raising the error as a standalone script 🙏
> 
> Thanks!

Thank you for your reply! I tried the run code from https://frgfm.github.io/torch-cam/cams.html#grad-cam, but failed. I got this error: "element 0 of tensors does not require grad and does not have a grad_fn". Then I tried to use the code from test_cams.py.  The  _forward function is come from your code torch-cam/test/test_cams.py  . But I still got this problem.

**Comment by sosoIrving** [comment]
> 
> 
> > Hi @sosoIrving,
> > Could you edit your issue description to display the error? I'm not sure whether this is a typo but the error you mentioned is not in it. If the error that you get is the one in the title, then it looks like the problem comes most likely from your forward. Gradient based CAM methods need to be able to backprop through the architecture. If you use for instance `with torch.no_grad()` then this will not work.
> > Also, I see that you use a function called `_forward` whose definition is not here. I need to be able to reproduce your error for me to be able to help 😅 Please ensure that the snippet you provide is raising the error as a standalone script 🙏
> > Thanks!
> 
> Thank you for your reply! I tried to run the code from https://frgfm.github.io/torch-cam/cams.html#grad-cam, but failed. I got this error: "element 0 of tensors does not require grad and does not have a grad_fn". Then I tried to use the code from test_cams.py. The _forward function is coming from your code torch-cam/test/test_cams.py . But I still got this problem.

**Comment by frgfm** [comment]
Hey there!

@lm-b yes, the logic is the following: `with torch.no_grad()` is a trick used by PyTorch to avoid tracking gradient while performing operations in that context. However, for some CAM, especially the GradCAM types, they do need gradient as stated in their name. But I agree that I should add a more informative error to help the user, I'll open a PR shortly :ok_hand: 

@sosoIrving oh my bad, the example has not been updated, I'll fix that right away. Please try my suggestion above :pray:

**Comment by frgfm** [comment]
I just merged the PR about the documentation, and thought about changing the error message. But actually, in my opinion, the error message is clear enough i.e. what the user is doing in the context cannot be checked here. 

Thus, I could only try the `.backward` and specify the exception. But this gradient issue could come from multiple reasons, so I don't think it's wise to change the message to something like "Consider removing torch.no_grad() context`. What do you think @lm-b @sosoIrving ?

**Comment by frgfm** [comment]
Closing this as I believe the issue is solved. Feel free to reopen if you disagree or encounter other related issues :ok_hand:
